lxd/storage_pools_config.go

```diff
@@ -0,0 +1,162 @@
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/lxc/lxd/shared"
+)
+
+var storagePoolConfigKeys = map[string]func(value string) error{
+	"source": shared.IsAny,
+	"size":   shared.IsAny,
+	"volume.block.mount_options":  shared.IsAny,
+	"volume.block.filesystem":     shared.IsAny,
+	"volume.size":                 shared.IsAny,
+	"volume.zfs.use_refquota":     shared.IsBool,
+	"volume.zfs.remove_snapshots": shared.IsBool,
+	"volume.lvm.thinpool_name":    shared.IsAny,
+	"zfs.pool_name":               shared.IsAny,
+}
+
+func storagePoolValidateConfig(name string, driver string, config map[string]string) error {
+	err := func(value string) error {
+		return shared.IsOneOf(value, supportedStorageTypes)
+	}(driver)
+	if err != nil {
+		return err
+	}
+
+	if config["source"] == "" {
+		if driver == "dir" {
+			config["source"] = filepath.Join(shared.VarPath("storage-pools"), name)
+		} else {
+			config["source"] = filepath.Join(shared.VarPath("disks"), name)
+		}
+	}
+
+	for key, val := range config {
+		// User keys are not validated.
+		if strings.HasPrefix(key, "user.") {
+			continue
+		}
+
+		// Validate storage pool config keys.
+		validator, ok := storagePoolConfigKeys[key]
+		if !ok {
+			return fmt.Errorf("Invalid storage pool configuration key: %s", key)
+		}
+
+		err := validator(val)
+		if err != nil {
+			return err
+		}
+
+		if driver != "zfs" || driver == "dir" {
+			if config["volume.zfs.use_refquota"] != "" {
+				return fmt.Errorf("Key volume.zfs.use_refquota cannot be used with non zfs storage pools.")
+			}
+
+			if config["volume.zfs.remove_snapshots"] != "" {
+				return fmt.Errorf("Key volume.zfs.remove_snapshots cannot be used with non zfs storage pools.")
+			}
+
+			if config["zfs.pool_name"] != "" {
+				return fmt.Errorf("Key zfs.pool_name cannot be used with non zfs storage pools.")
+			}
+		}
+
+		if driver == "dir" {
+			if config["size"] != "" {
+				return fmt.Errorf("Key size cannot be used with dir storage pools.")
+			}
+
+			if config["volume.block.mount_options"] != "" {
+				return fmt.Errorf("Key volume.block.mount_options cannot be used with dir storage pools.")
+			}
+
+			if config["volume.block.filesystem"] != "" {
+				return fmt.Errorf("Key volume.block.filesystem cannot be used with dir storage pools.")
+			}
+
+			if config["volume.size"] != "" {
+				return fmt.Errorf("Key volume.size cannot be used with dir storage pools.")
+			}
+		}
+	}
```

Comment:
All of the above errors read a bit funny, especially the "Key size" one. I think they should be changed to:

"The \"XYZ\" key cannot be used with ..."